midp/fs.js

```diff
@@ -267,375 +263,463 @@ Native["com/sun/midp/rms/RecordStoreRegistry.stopAllRecordStoreListeners.(I)V"]
-    files = files.filter(regexp.test.bind(regexp));
-    var filesArray = J2ME.newArray(J2ME.PrimitiveArrayClassInfo.B.klass, files.length);
-    var encoder = new TextEncoder("utf-8");
 
-    files.forEach(function(file, i) {
-        var bytesFile = encoder.encode(file);
-        var fileArray = J2ME.newByteArray(bytesFile.byteLength);
-        fileArray.set(bytesFile);
-        filesArray[i] = fileArray;
-    });
+Native["com/sun/cdc/io/j2me/file/DefaultFileHandler.rename0.(Ljava/lang/String;)V"] = function(newName) {
+    var pathname = util.fromJavaString(this.$nativePath);
+    var newPathname = util.fromJavaString(newName);
+    DEBUG_FS && console.log("DefaultFileHandler.rename0: " + pathname + " to " + newPathname);
 
-    return filesArray;
-};
+    if (fs.exists(newPathname)) {
+        throw $.newIOException("file with new name exists");
+    }
 
-Native["com/ibm/oti/connection/file/Connection.mkdirImpl.([B)I"] = function(path) {
-    // IBM's implementation returns different error numbers, we don't care
-    return fs.mkdir(getAbsolutePath(path)) ? 0 : 42;
+    if (!fs.rename(pathname, newPathname)) {
+        throw $.newIOException("error renaming file");
+    }
 };
 
-Native["com/ibm/oti/connection/file/Connection.newFileImpl.([B)I"] = function(jPath) {
-    var path = getAbsolutePath(jPath);
+Native["com/sun/cdc/io/j2me/file/DefaultFileHandler.truncate.(J)V"] = function(byteOffset) {
+    var pathname = util.fromJavaString(this.$nativePath);
+    DEBUG_FS && console.log("DefaultFileHandler.lastModified: " + pathname);
+
+    var stat = fs.stat(pathname);
 
-    var stat = fs.stat(path);
-    if (stat !== null) {
-        return stat.isDir ? 3 : 1;
+    if (!stat) {
+        throw $.newIOException("file does not exist");
     }
 
-    return fs.create(path, new Blob()) ? 0 : 42;
+    if (stat.isDir) {
+        throw $.newIOException("file is directory");
+    }
+
+    // TODO: If the file is open, flush it first.
+
+    fs.truncate(pathname, byteOffset.toNumber());
 };
 
-Native["com/ibm/oti/connection/file/Connection.deleteFileImpl.([B)Z"] = function(path) {
-    return fs.remove(getAbsolutePath(path)) ? 1 : 0;
+Native["com/sun/cdc/io/j2me/file/DefaultFileHandler.fileSize.()J"] = function() {
+    var pathname = util.fromJavaString(this.$nativePath);
+    DEBUG_FS && console.log("DefaultFileHandler.fileSize: " + pathname);
+
+    return Long.fromNumber(fs.size(pathname));
 };
 
-Native["com/ibm/oti/connection/file/Connection.deleteDirImpl.([B)Z"] =
-  Native["com/ibm/oti/connection/file/Connection.deleteFileImpl.([B)Z"];
+addUnimplementedNative("com/sun/cdc/io/j2me/file/DefaultFileHandler.directorySize.(Z)J", Long.fromNumber(0));
 
-Native["com/ibm/oti/connection/file/Connection.isReadOnlyImpl.([B)Z"] = function(path) {
-    console.warn("Connection.isReadOnlyImpl.([B)Z not implemented (" + getAbsolutePath(path) + ")");
-    return 0;
+Native["com/sun/cdc/io/j2me/file/DefaultFileHandler.canRead.()Z"] = function() {
```

Comment:
> We can probably avoid checking if the file exists here, looks like the Java implementation always checks if the file exists before calling canRead (the same applies to canWrite)

*Protocol.openInputStream* checks *DefaultFileHandler.exists* before it calls *DefaultFileHandler.canRead*, but *Protocol.canRead* doesn't do that, and that method is public (an implementation of *FileConnection.canRead*), so it could be called by midlets.